rate-limiting-global/src/main/java/io/stargate/db/limiter/global/impl/GlobalRateLimitingManager.java

```diff
@@ -0,0 +1,143 @@
+      throw new IllegalArgumentException(
+          format(
+              "Global rate limiting is enabled but missing (or empty) value for property '%s'",
+              RATE_PROPERTY));
+    }
+
+    try {
+      return new AsyncRateLimiter(
+          executor, Long.parseLong(rateStr), TimeUnit.SECONDS, 1, TimeUnit.MINUTES);
+    } catch (NumberFormatException e) {
+      throw new IllegalArgumentException(
+          format(
+              "Invalid value for property '%s': " + "expected a number, but got %s",
+              RATE_PROPERTY, rateStr));
+    }
+  }
+
+  @Override
+  public String description() {
+    return format("global rate limiting at %d queries/seconds", limiter.getRate(TimeUnit.SECONDS));
+  }
+
+  @Override
+  public ConnectionManager forNewConnection() {
+    // Nothing is specific to connection for this rate limiter
+    return this;
+  }
+
+  @Override
+  public ConnectionManager forNewConnection(ClientInfo clientInfo) {
+    // Nothing is specific to connection for this rate limiter
+    return this;
+  }
+
+  @Override
+  public void onUserLogged(AuthenticatedUser user) {
+    // Nothing specific to users.
+  }
+
+  private static boolean shouldExclude(String query) {
+    // TODO: this is a horrible horrible hack. As we only pass strings through `Persistence`, we
+    //   have no simple reliable way to know the keyspace or table queried. But if we want to use
+    //   this for testing somewhat reliably, we need to be able to exclude the bunch of system table
+    //   reads that drivers do on connection (basically, for testing to be reliable, we want to set
+    //   a  very aggressive rate limiting, but if driver queries hit it, the test will take
+    // forever).
+    //   So we explicitly target those driver system table reads. It's super fragile because 1) it
+    //   assume the driver don't add extra spaces compared to what we check and 2) it doesn't use
+    //   prepared queries (if it did, we wouldn't rate limit the prepare, but we would the
+    //   execution).
+    //   Tl;dr: we'd need Stargate to stop passing raw strings into the persistence to be able to
+    //   do this properly.
+    return RATE_LIMIT_SYSTEM_TABLES && query.startsWith("SELECT") && query.contains("FROM System");
+  }
+
+  @Override
+  public RateLimitingDecision forPrepare(String query, Parameters parameters) {
+    if (shouldExclude(query)) {
+      return RateLimitingDecision.unlimited();
+    }
+    return RateLimitingDecision.limit(limiter, 1);
+  }
+
+  @Override
+  public RateLimitingDecision forExecute(Statement statement, Parameters parameters) {
+    if (statement instanceof SimpleStatement
+        && shouldExclude(((SimpleStatement) statement).queryString())) {
+      return RateLimitingDecision.unlimited();
+    }
+    return RateLimitingDecision.limit(limiter, 1);
+  }
+
+  @Override
+  public RateLimitingDecision forBatch(Batch batch, Parameters parameters) {
+    return RateLimitingDecision.limit(limiter, 1);
```

Comment:
Why do we acquire 1 permit here but not `batch.size()`? I assume a permit corresponds to one "basic" query, so would it make sense to make batch cost progressively higher according to the number of queries it contains?

Reply:
Maybe. I could argue that the rate limiting is not really based on the queries "size" in that a full-table scan will be rate limited the same than a 1 row query, so it's not that clear that making a special case for batch is decidedly better or worst. But I changed it, because both are meaningful options and I don't really care. As noted in the comments in the code, I added that `GlobalRateLimitingManager` to have a real example and to have a semi-realistic test, but I don't really believe this is for using in production in this state so ....